comtypes/test/test_inout_args.py

```diff
@@ -0,0 +1,485 @@
+        self.assertEqual(orig_kw["pdwOptimalWriteBufferSize"].value, buf_size)
+        self.assertIsInstance(orig_kw["ppszCookie"], WSTRING)
+        self.assertEqual(orig_kw["ppszCookie"].value, cookie)
+
+    def test_mixed_args_1(self):
+        self_ = MagicMock(name="Self")
+        p_val = MagicMock(spec=POINTER(IUnknown))()
+        buf_size = 3
+        cookie = "h"
+        ret_val = self.fixed(
+            self_, p_val, ppszCookie=cookie, pdwOptimalWriteBufferSize=buf_size
+        )
+
+        self.assertEqual(ret_val, [self.pp_data, buf_size, cookie])
+        self.orig.assert_called_once()
+        (orig_0th, orig_1st), orig_kw = self.orig.call_args
+        self.assertIs(orig_0th, self_)
+        self.assertEqual(orig_1st, p_val)
+        self.assertEqual(set(orig_kw), {"pdwOptimalWriteBufferSize", "ppszCookie"})
+        self.assertIsInstance(orig_kw["pdwOptimalWriteBufferSize"], c_ulong)
+        self.assertEqual(orig_kw["pdwOptimalWriteBufferSize"].value, buf_size)
+        self.assertIsInstance(orig_kw["ppszCookie"], WSTRING)
+        self.assertEqual(orig_kw["ppszCookie"].value, cookie)
+
+    def test_mixed_args_2(self):
+        self_ = MagicMock(name="Self")
+        p_val = MagicMock(spec=POINTER(IUnknown))()
+        buf_size = 2
+        cookie = "ij"
+        ret_val = self.fixed(self_, p_val, buf_size, ppszCookie=cookie)
+
+        self.assertEqual(ret_val, [self.pp_data, buf_size, cookie])
+        self.orig.assert_called_once()
+        (orig_0th, orig_1st, orig_2nd), orig_kw = self.orig.call_args
+        self.assertIs(orig_0th, self_)
+        self.assertEqual(orig_1st, p_val)
+        self.assertEqual(set(orig_kw), {"ppszCookie"})
+        self.assertIsInstance(orig_2nd, c_ulong)
+        self.assertEqual(orig_2nd.value, buf_size)
+        self.assertIsInstance(orig_kw["ppszCookie"], WSTRING)
+        self.assertEqual(orig_kw["ppszCookie"].value, cookie)
+
+    def test_omitted_arguments_autogen(self):
+        self_ = MagicMock(name="Self")
+        p_val = MagicMock(spec=POINTER(IUnknown))()
+        ret_val = self.fixed(self_, pValues=p_val)
+
+        self.orig.assert_called_once()
+        (orig_0th,), orig_kw = self.orig.call_args
+        self.assertEqual(
+            set(orig_kw), {"pValues", "pdwOptimalWriteBufferSize", "ppszCookie"}
+        )
+        self.assertIs(orig_0th, self_)
+        self.assertIs(orig_kw["pValues"], p_val)
+        self.assertIsInstance(orig_kw["pdwOptimalWriteBufferSize"], c_ulong)
+        self.assertEqual(orig_kw["pdwOptimalWriteBufferSize"].value, c_ulong().value)
+        self.assertIsInstance(orig_kw["ppszCookie"], WSTRING)
+        self.assertEqual(orig_kw["ppszCookie"].value, WSTRING().value)
+
+        self.assertEqual(
+            ret_val,
+            [
+                self.pp_data,
+                orig_kw["pdwOptimalWriteBufferSize"].value,
+                orig_kw["ppszCookie"].value,
+            ],
+        )
+
```

Comment:
```suggestion


@ut.expectedFailure
class Test_ArgsKwargsCombinations_2(ut.TestCase):
    def setUp(self):
        # a memberspec of `PortableDeviceApiLib.IPortableDeviceContent`
        spec = comtypes.COMMETHOD(
            [],
            HRESULT,
            "CreateObjectWithPropertiesOnly",
            (["in"], POINTER(comtypes.IUnknown), "pValues"),  # IPortableDeviceValues
            (["in", "out"], POINTER(WSTRING), "ppszObjectID"),
        )
        self.orig = MagicMock()
        self.fixed = _fix_inout_args(self.orig, spec.argtypes, spec.paramflags)

    def test_positionals_only(self):
        self_ = MagicMock(name="Self")
        p_val = MagicMock(spec=POINTER(IUnknown))()
        objid = "foo"
        ret_val = self.fixed(self_, p_val, objid)

        self.assertEqual(ret_val, objid)
        self.orig.assert_called_once()
        (orig_0th, orig_1st, orig_2nd), orig_kw = self.orig.call_args
        self.assertIs(orig_0th, self_)
        self.assertIs(orig_1st, p_val)
        self.assertIsInstance(orig_2nd, WSTRING)
        self.assertEqual(orig_2nd.value, objid)
        self.assertEqual(orig_kw, {})

    def test_keywords_only(self):
        self_ = MagicMock(name="Self")
        p_val = MagicMock(spec=POINTER(IUnknown))()
        objid = "bar"
        ret_val = self.fixed(self_, ppszObjectID=objid, pValues=p_val)

        self.assertEqual(ret_val, objid)
        self.orig.assert_called_once()
        (orig_0th,), orig_kw = self.orig.call_args
        self.assertIs(orig_0th, self_)
        self.assertIs(orig_kw["pValues"], p_val)
        self.assertIsInstance(orig_kw["ppszObjectID"], WSTRING)
        self.assertEqual(orig_kw["ppszObjectID"].value, objid)

    def test_mixed_args(self):
        self_ = MagicMock(name="Self")
        p_val = MagicMock(spec=POINTER(IUnknown))()
        objid = "baz"
        ret_val = self.fixed(self_, p_val, ppszObjectID=objid)

        self.assertEqual(ret_val, objid)
        self.orig.assert_called_once()
        (orig_0th, orig_1st), orig_kw = self.orig.call_args
        self.assertIs(orig_0th, self_)
        self.assertIs(orig_1st, p_val)
        self.assertIsInstance(orig_kw["ppszObjectID"], WSTRING)
        self.assertEqual(orig_kw["ppszObjectID"].value, objid)

    def test_omitted_arguments_autogen(self):
        self_ = MagicMock(name="Self")
        p_val = MagicMock(spec=POINTER(IUnknown))()
        self.orig.return_value = "qux"
        ret_val = self.fixed(self_, pValues=p_val)

        self.orig.assert_called_once()
        (orig_0th,), orig_kw = self.orig.call_args
        self.assertEqual(ret_val, orig_kw["ppszObjectID"].value)
        self.assertIs(orig_0th, self_)
        self.assertEqual(set(orig_kw), {"pValues", "ppszObjectID"})
        self.assertIs(orig_kw["pValues"], p_val)
        self.assertIsInstance(orig_kw["ppszObjectID"], WSTRING)
        self.assertEqual(orig_kw["ppszObjectID"].value, WSTRING().value)

```
I came up with a new test and wrote it, but looking back after you added the new test, this is not `real_world_examples`, but rather a test that covers the case where there is `['in', 'out']` but `['out']` is not in the memberspec . And this test will not work without resolving the problem that you explained below, so this testcase is just suggestion.

https://github.com/enthought/comtypes/blob/3b72578ff58af2b1e984304cfac1e7bf41fd1702/comtypes/_memberspec.py#L238-L257

I understood clearly your explanation after writing my own test code.
Indeed, It seems to me this is a wierd behavior.